service-workers/service-worker/ready.https.html

```diff
@@ -250,51 +250,68 @@
-    // with an active promise for the matching scope.
-    await Promise.race([readyPromise, timeoutPromise]);
-
-    assert_not_equals(timeoutId, null,
-                      'ready promise should resolve before timeout');
-    clearTimeout(timeoutId);
-
-    // We should get here and not time out.
-
-  }, 'access ready on uninstalling registration that is resurrected');
+  const url1 = 'resources/empty-worker.js';
+  const url2 = url1 + '?2';
+  const matched_scope = 'resources/blank.html?ready-after-unregister';
+
+  const reg1 = await service_worker_unregister_and_register(t, url1, matched_scope);
+  t.add_cleanup(() => reg1.unregister());
+
+  await wait_for_state(t, reg1.installing, 'activating');
+  // This registration will resolve all ready promises in clients that match the scope.
+  // But there are no clients.
+
+  const frame = await with_iframe(matched_scope);
+  t.add_cleanup(() => frame.remove());
+
+  await reg1.unregister();
+
+  // Access the ready promise while the registration is unregistering.
+  const readyPromise = frame.contentWindow.navigator.serviceWorker.ready;
+
+  // Create a new registration.
+  const reg2 = await navigator.serviceWorker.register(url2, { scope: matched_scope });
+  t.add_cleanup(() => reg2.unregister());
+  // This registration will resolve all ready promises in clients that match the scope.
+  // That includes frame's client.
+
+  const readyReg = await readyPromise;
+
+  assert_equals(readyReg.active.scriptURL, reg2.active.scriptURL, 'Resolves with the second registration');
+  assert_not_equals(reg1, reg2, 'Registrations should be different');
+}, 'resolve ready after unregistering and reregistering');
+
+promise_test(async function(t) {
+  const url1 = 'resources/empty-worker.js';
+  const url2 = url1 + '?2';
+  const matched_scope = 'resources/blank.html?ready-after-unregister';
+
+  const frame = await with_iframe(matched_scope);
+  t.add_cleanup(() => frame.remove());
+
+  const reg1 = await service_worker_unregister_and_register(t, url1, matched_scope);
+  t.add_cleanup(() => reg1.unregister());
+
+  await wait_for_state(t, reg1.installing, 'activated');
+  // This registration will resolve all ready promises in clients that match the scope.
+  // That includes frame's client.
+
+  const reg1Active = reg1.active;
+
+  await reg1.unregister();
+
+  // Access the ready promise while the registration is unregistering.
+  const readyPromise = frame.contentWindow.navigator.serviceWorker.ready;
+
+  // Create a new registration.
+  const reg2 = await navigator.serviceWorker.register(url2, { scope: matched_scope });
+  t.add_cleanup(() => reg2.unregister());
+  // This registration will resolve all ready promises in clients that match the scope.
+  // That includes frame's client, but its ready promise has already resolved.
+
+  const readyReg = await readyPromise;
+
+  assert_equals(readyReg.active.scriptURL, reg1Active.scriptURL, 'Resolves with the first registration');
```

Comment:
@jakearchibald Could you explain where this assumption is coming from?
The way I'd interpret the spec here is that it should be the url of the new worker, which is also how Firefox seems to understand it [1].

In step 4 of the `Unregister algorithm` [2], we remove the registration from `scope to registration map`. Step 2.1 of `serviceWorker.ready` [3] invokes the `Match Service Worker Registration algorithm` [4], which only looks at registrations that are in `scope to registration map`. Since there is no non-uninstalled registration at the time we're calling `serviceWorker.ready`, the Promise should only resolve after the new worker is registered. Hence, we shouln't get the registration we intent to assert against here.

Am I missing something, or is this a bug in the test?

//cc @mattto 

[1] https://wpt.fyi/results/service-workers/service-worker/ready.https.html?label=master&label=experimental&aligned&q=service-workers%2Fservice-worker%2Fready.https.html
[2] https://w3c.github.io/ServiceWorker/#unregister-algorithm
[3] https://w3c.github.io/ServiceWorker/#navigator-service-worker-ready
[4] https://w3c.github.io/ServiceWorker/#match-service-worker-registration

Reply:
Won't the promise be resolved earlier, in step 7.2 of https://w3c.github.io/ServiceWorker/#activation-algorithm?

Reply:
That'd be true if we swap lines 301 and 304.

`.ready` returns a new Promise every time it's called and runs https://w3c.github.io/ServiceWorker/#match-service-worker-registration to resolve it. Resolving it once after activating and return the same Promise for every call would contradict `ready after a longer matched registration registered` in this file.

Reply:
> `.ready` returns a new Promise every time it's called

That isn't how I interpret the spec. [A service worker container has a ready promise](https://w3c.github.io/ServiceWorker/#serviceworkercontainer-ready-promise), so the promise is created the same time the service worker container is created.

Reply:
You're right, that's what the spec says. Thanks for the explanation!

With that model, how do we explain that `.ready` resolves to the second worker [here](https://github.com/web-platform-tests/wpt/blob/84022abe9d31160f4b09e3d146479bb74f73e097/service-workers/service-worker/ready.https.html#L252) and [here](https://github.com/web-platform-tests/wpt/blob/84022abe9d31160f4b09e3d146479bb74f73e097/service-workers/service-worker/ready.https.html#L166)? Shouln't it be resolved by the first activation?

Reply:
Oof this isn't how Chrome implements .ready. Chrome doesn't try to resolve the promise until .ready is accessed. Then once it resolves, it always resolves with the same registration. I agree though the spec says to resolve each client's ready promise at activation time.

For the first link, the test matches the spec because `frame` is created after the activation of the first worker.

For the second link, I think you're right. `frame` is created, then the first activation occurs, so .ready should resolve to the first registration.

If browsers are all doing the "don't try to figure the value until the property is accessed thing", we could potentially change the spec for this.

Reply:
Filed https://github.com/w3c/ServiceWorker/issues/1477

Reply:
I feel like the spec used to match the implementations.  For example, I reference this behavior in this issue:

https://github.com/w3c/ServiceWorker/issues/1278#issuecomment-364580477